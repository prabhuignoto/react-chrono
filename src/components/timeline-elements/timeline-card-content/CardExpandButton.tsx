import React from 'react';
import cls from 'classnames';
import ChevronDown from '../../icons/chev-down';
import { showMoreButton, chevronIconWrapper } from './timeline-card-content.css';

interface CardExpandButtonProps {
  expanded: boolean;
  onToggle: () => void;
  expandable: boolean;
  showMore: string | undefined;
  showLess: string | undefined;
}

export const CardExpandButton: React.FC<CardExpandButtonProps> = React.memo(({
  expanded,
  onToggle,
  expandable,
  showMore = 'Show More',
  showLess = 'Show Less'
}) => {
  if (!expandable) return null;

  return (
    <button
      className={showMoreButton}
      onClick={onToggle}
      aria-expanded={expanded}
      aria-label={expanded ? showLess : showMore}
      type="button"
    >
      {expanded ? showLess : showMore}
      <span 
        className={cls(chevronIconWrapper, {
          expanded: expanded
        })}
      >
        <ChevronDown />
      </span>
    </button>
  );
});