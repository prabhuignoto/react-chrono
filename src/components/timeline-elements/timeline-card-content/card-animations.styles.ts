import { keyframes } from 'styled-components';

export const reveal = keyframes`
  0% {
    opacity: 0;
    transform: scale(0.95);
  }
  100% {
    opacity: 1;
    transform: scale(1);
  }
`;

export const slideInFromTop = keyframes`
  0% {
    opacity: 0;
    transform: translateY(-50%);
  }
  100% {
    opacity: 1;
    transform: translateY(0);
  }
`;

export const slideInFromLeft = keyframes`
  0% {
    opacity: 0;
    transform: translateX(-50%);
  }
  100% {
    opacity: 1;
    transform: translateX(0);
  }
`;

export const slideFromRight = keyframes`
  0% {
    opacity: 0;
    transform: translateX(50%);
  }
  100% {
    opacity: 1;
    transform: translateX(0);
  }
`;
