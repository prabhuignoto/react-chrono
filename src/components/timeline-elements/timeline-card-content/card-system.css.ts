import { style, globalStyle } from '@vanilla-extract/css';
import { recipe } from '@vanilla-extract/recipes';
import { designTokens } from '../../../styles/design-system.css';
import { animationUtils, transitions } from '../../../styles/animations.css';
import { layoutBase } from '../../../styles/layout.css';

// Base card foundation
const cardBase = style([
  {
    background: `linear-gradient(135deg, ${designTokens.color.card.bg} 0%, ${designTokens.color.card.bg}f5 100%)`,
    borderRadius: designTokens.radius.xl,
    border: `1px solid ${designTokens.color.card.border}`,
    boxShadow: designTokens.elevation.md,
    backdropFilter: 'blur(20px)',
    transition: transitions.normal,
    position: 'relative',
    overflow: 'hidden',
  },
]);

// Card recipe with variants
export const card = recipe({
  base: [cardBase, animationUtils.respectsReducedMotion],

  variants: {
    size: {
      sm: {
        minWidth: '200px',
        padding: designTokens.spacing.sm,
        borderRadius: designTokens.radius.md,
      },
      md: {
        minWidth: '280px',
        padding: designTokens.spacing.md,
      },
      lg: {
        minWidth: '320px',
        padding: designTokens.spacing.lg,
      },
    },

    elevation: {
      flat: { boxShadow: designTokens.elevation.none },
      low: { boxShadow: designTokens.elevation.sm },
      medium: { boxShadow: designTokens.elevation.md },
      high: { boxShadow: designTokens.elevation.lg },
    },

    interactive: {
      true: {
        cursor: 'pointer',
        selectors: {
          '&:hover': {
            transform: 'translateY(-2px)',
            boxShadow: designTokens.elevation.xl,
          },
          '&:active': {
            transform: 'translateY(0px)',
            transition: transitions.fast,
          },
        },
      },
      false: {},
    },

    animation: {
      none: {},
      fadeIn: animationUtils.fadeIn,
      slideLeft: animationUtils.entranceLeft,
      slideRight: animationUtils.entranceRight,
      slideTop: animationUtils.entranceTop,
    },

    bordered: {
      true: {
        '::before': {
          content: '""',
          position: 'absolute',
          inset: 0,
          borderRadius: 'inherit',
          padding: '1px',
          background: `linear-gradient(135deg, ${designTokens.color.primary}20 0%, transparent 50%, ${designTokens.color.primary}10 100%)`,
          mask: 'linear-gradient(#fff 0 0) content-box, linear-gradient(#fff 0 0)',
          maskComposite: 'xor',
          WebkitMask:
            'linear-gradient(#fff 0 0) content-box, linear-gradient(#fff 0 0)',
          WebkitMaskComposite: 'xor',
          pointerEvents: 'none',
        },
      },
      false: {},
    },
  },

  defaultVariants: {
    size: 'md',
    elevation: 'medium',
    interactive: false,
    animation: 'none',
    bordered: false,
  },

  compoundVariants: [
    // Mobile optimizations
    {
      variants: { size: 'lg' },
      style: {
        '@media': {
          '(max-width: 768px)': {
            minWidth: '240px',
          },
          '(max-width: 480px)': {
            minWidth: '200px',
          },
        },
      },
    },
  ],
});

// Card content structure
export const cardContent = recipe({
  base: [
    layoutBase.flexColumn,
    {
      gap: designTokens.spacing.sm,
      width: '100%',
    },
  ],

  variants: {
    spacing: {
      tight: { gap: designTokens.spacing.xs },
      normal: { gap: designTokens.spacing.sm },
      loose: { gap: designTokens.spacing.md },
    },

    alignment: {
      start: { alignItems: 'flex-start' },
      center: { alignItems: 'center' },
      end: { alignItems: 'flex-end' },
    },
  },

  defaultVariants: {
    spacing: 'normal',
    alignment: 'start',
  },
});

// Card header
export const cardHeader = style([
  layoutBase.flexColumn,
  {
    gap: designTokens.spacing.xs,
    marginBottom: designTokens.spacing.sm,
  },
]);

// Card title
export const cardTitle = style({
  fontSize: designTokens.typography.fontSize.lg,
  fontWeight: designTokens.typography.fontWeight.semibold,
  lineHeight: designTokens.typography.lineHeight.tight,
  color: designTokens.color.card.title,
  margin: 0,
});

// Card subtitle
export const cardSubtitle = style({
  fontSize: designTokens.typography.fontSize.sm,
  fontWeight: designTokens.typography.fontWeight.medium,
  lineHeight: designTokens.typography.lineHeight.normal,
  color: designTokens.color.card.subtitle,
  margin: 0,
  opacity: 0.8,
});

// Card body
export const cardBody = style({
  fontSize: designTokens.typography.fontSize.base,
  lineHeight: designTokens.typography.lineHeight.relaxed,
  color: designTokens.color.card.details,
});

// Focus management for interactive cards
globalStyle(`${card.classNames.base}:focus-visible`, {
  outline: `2px solid ${designTokens.color.interactive.focusRing}`,
  outlineOffset: '2px',
});
