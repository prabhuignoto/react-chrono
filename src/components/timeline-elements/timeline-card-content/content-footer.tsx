import React, {
  FunctionComponent,
  RefObject,
  useContext,
  useMemo,
} from 'react';
import { Theme } from '../../../models/Theme';
import { TimelineMode } from '../../../models/TimelineModel';
import { GlobalContext } from '../../GlobalContext';
import ChevronIcon from '../../icons/chev-right';
import {
  ChevronIconWrapper,
  ShowMore,
  SlideShowProgressBar,
  TriangleIconWrapper,
} from './timeline-card-content.styles';

export type ContentFooterProps = {
  canShow: boolean;
  onExpand: () => void;
  paused: boolean;
  progressRef: RefObject<HTMLDivElement>;
  remainInterval: number;
  showMore: boolean;
  showProgressBar?: boolean;
  showReadMore?: boolean | '';
  startWidth: number;
  textContentIsLarge: boolean;
  theme?: Theme;
  triangleDir?: string;
};

const ContentFooter: FunctionComponent<ContentFooterProps> = ({
  theme,
  showProgressBar,
  onExpand,
  triangleDir,
  showMore,
  textContentIsLarge,
  showReadMore,
  remainInterval,
  paused,
  startWidth,
  canShow,
  progressRef,
}) => {
  const { mode } = useContext(GlobalContext);

  const canShowTriangleIcon = useMemo(() => {
    return (['VERTICAL', 'VERTICAL_ALTERNATING'] as TimelineMode[]).some(
      (m) => m === mode,
    );
  }, [mode]);

  return (
    <>
      {showReadMore && textContentIsLarge ? (
        <ShowMore
          className="show-more"
          onClick={onExpand}
          onKeyPress={(event) => {
            if (event.key === 'Enter') {
              onExpand();
            }
          }}
          role="button"
          show={canShow}
          theme={theme}
          tabIndex={0}
        >
          {<span>{showMore ? 'read less' : 'read more'}</span>}
          <ChevronIconWrapper collapsed={!showMore}>
            <ChevronIcon />
          </ChevronIconWrapper>
        </ShowMore>
      ) : null}

      {showProgressBar && (
        <SlideShowProgressBar
          color={theme?.primary}
          duration={remainInterval}
          paused={paused}
          ref={progressRef}
          startWidth={startWidth}
        ></SlideShowProgressBar>
      )}

      {canShowTriangleIcon && (
        <TriangleIconWrapper
          dir={triangleDir}
          theme={theme}
        ></TriangleIconWrapper>
      )}
    </>
  );
};

export { ContentFooter };
