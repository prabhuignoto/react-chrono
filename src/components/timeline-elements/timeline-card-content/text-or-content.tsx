import { TimelineContentModel } from '@models/TimelineContentModel';
import { TimelineProps } from '@models/TimelineModel';
import {
  ForwardRefExoticComponent,
  ReactNode,
  forwardRef,
  useContext,
  useMemo,
} from 'react';
import xss from 'xss';
import { GlobalContext } from '../../GlobalContext';
import {
  TimelineContentDetails,
  TimelineSubContent,
} from './timeline-card-content.styles';

// Define the type for the TextOrContentModel
export type TextOrContentModel = Pick<
  TimelineContentModel,
  'timelineContent' | 'theme' | 'detailedText'
> & {
  showMore?: boolean;
};

// Function to render an array of text
const renderTextArray: (
  p: Pick<TimelineProps, 'parseDetailsAsHTML' | 'fontSizes' | 'theme'> & {
    cardTextClassName: string;
    detailedText: string[];
  },
) => ReactNode = ({
  fontSizes,
  parseDetailsAsHTML,
  theme,
  detailedText,
  cardTextClassName,
}) => {
  return detailedText.map((text, index) => {
    const props = parseDetailsAsHTML
      ? {
          dangerouslySetInnerHTML: {
            __html: xss(text),
          },
        }
      : null;
    return (
      <TimelineSubContent
        key={index}
        fontSize={fontSizes?.cardText}
        className={cardTextClassName}
        theme={theme}
        {...props}
      >
        {parseDetailsAsHTML ? null : text}
      </TimelineSubContent>
    );
  });
};

// Function to get the TextOrContent component
const getTextOrContent: (
  p: TextOrContentModel,
) => ForwardRefExoticComponent<TextOrContentModel> = ({
  timelineContent,
  theme,
  detailedText,
  showMore,
}) => {
  const TextOrContent = forwardRef<HTMLParagraphElement, TextOrContentModel>(
    (prop, ref) => {
      const isTextArray = Array.isArray(detailedText);
      const { fontSizes, classNames, parseDetailsAsHTML, textDensity } =
        useContext(GlobalContext);

      const renderTimelineContent = () => {
        if (timelineContent) {
          return <div ref={ref}>{timelineContent}</div>;
        } else {
          let textContent = null;
          if (isTextArray) {
            textContent = renderTextArray({
              cardTextClassName: classNames?.cardText,
              detailedText,
              fontSizes,
              parseDetailsAsHTML,
              theme,
            });
          } else {
            textContent = parseDetailsAsHTML ? xss(detailedText) : detailedText;
          }

          const textContentProps =
            parseDetailsAsHTML && !isTextArray
              ? {
                  dangerouslySetInnerHTML: {
                    __html: xss(textContent),
                  },
                }
              : {};

          const shouldNotShowText = useMemo(() => {
            return (
              (parseDetailsAsHTML && !isTextArray) || textDensity === 'LOW'
            );
          }, [isTextArray, textDensity]);

          return textContent ? (
            <TimelineContentDetails
              className={showMore ? 'active' : ''}
              ref={ref}
              theme={theme}
              {...textContentProps}
            >
              {shouldNotShowText ? null : textContent}
            </TimelineContentDetails>
          ) : null;
        }
      };

      return renderTimelineContent();
    },
  );

  TextOrContent.displayName = 'Text Or Content';

  return TextOrContent;
};

export { getTextOrContent };
