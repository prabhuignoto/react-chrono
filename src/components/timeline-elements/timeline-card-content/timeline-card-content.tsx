import cls from 'classnames';
import React, {
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import { TimelineContentModel } from '../../../models/TimelineContentModel';
import { MediaState } from '../../../models/TimelineMediaModel';
import { TimelineMode } from '../../../models/TimelineModel';
import { GlobalContext } from '../../GlobalContext';
import ChevronIcon from '../../icons/chev-right';
import { MemoSubTitle, MemoTitle } from '../memoized';
import CardMedia from '../timeline-card-media/timeline-card-media';
import {
  ChevronIconWrapper,
  ShowMore,
  SlideShowProgressBar,
  TimelineCardHeader,
  TimelineContentDetails,
  TimelineContentDetailsWrapper,
  TimelineItemContentWrapper,
  TimelineSubContent,
  TriangleIconWrapper,
} from './timeline-card-content.styles';

const TimelineCardContent: React.FunctionComponent<TimelineContentModel> =
  React.memo(
    ({
      active,
      content,
      detailedText,
      id,
      media,
      onShowMore,
      slideShowActive,
      onElapsed,
      theme,
      title,
      onClick,
      customContent,
      hasFocus,
      flip,
      branchDir,
      url,
    }: TimelineContentModel) => {
      const [showMore, setShowMore] = useState(false);
      const detailsRef = useRef<HTMLDivElement | null>(null);
      const containerRef = useRef<HTMLDivElement | null>(null);
      const progressRef = useRef<HTMLDivElement | null>(null);

      const containerWidth = useRef<number>(0);
      const slideShowElapsed = useRef(0);
      const timerRef = useRef(0);
      const startTime = useRef<Date>();
      const [paused, setPaused] = useState(false);

      // const [elapsed, setElapsed] = useState(0);
      const [remainInterval, setRemainInterval] = useState(0);
      const [startWidth, setStartWidth] = useState(0);
      const [textContentLarge, setTextContentLarge] = useState(false);

      const {
        mode,
        cardHeight,
        slideItemDuration = 2000,
        useReadMore,
        cardWidth,
        borderLessCards,
        disableAutoScrollOnClick,
      } = useContext(GlobalContext);

      const canShowProgressBar = useMemo(() => {
        const canShow = active && slideShowActive;
        if (media) {
          return canShow && media.type !== 'VIDEO';
        } else {
          return canShow;
        }
      }, [active, slideShowActive]);

      const canShowMore = useMemo(() => {
        return !!detailedText;
      }, [detailedText]);

      useEffect(() => {
        const detailsEle = detailsRef.current;

        if (detailsEle) {
          detailsEle.scrollTop = 0;
        }
      }, [showMore]);

      const onContainerRef = useCallback(
        (node) => {
          const detailsEle = detailsRef.current;
          if (node && detailsEle) {
            const { scrollHeight, offsetTop } = detailsEle;
            containerWidth.current = node.clientWidth;
            setStartWidth(containerWidth.current);
            setTextContentLarge(scrollHeight + offsetTop > node.clientHeight);
          }
        },
        [detailsRef.current],
      );

      const setupTimer = useCallback((interval: number) => {
        if (!slideItemDuration) {
          return;
        }

        setRemainInterval(interval);

        startTime.current = new Date();

        setPaused(false);

        timerRef.current = window.setTimeout(() => {
          // clear the timer and move to the next card
          window.clearTimeout(timerRef.current);
          setPaused(true);
          setStartWidth(0);
          setRemainInterval(slideItemDuration);
          id && onElapsed && onElapsed(id);
        }, interval);
      }, []);

      // pause the slide show
      const tryHandlePauseSlideshow = useCallback(() => {
        if (active && slideShowActive) {
          window.clearTimeout(timerRef.current);
          setPaused(true);

          if (startTime.current) {
            const elapsed: any = +new Date() - +startTime.current;
            slideShowElapsed.current = elapsed;
          }

          if (progressRef.current) {
            setStartWidth(progressRef.current.clientWidth);
          }
        }
      }, [active, slideShowActive]);

      // resumes the slide show
      const tryHandleResumeSlideshow = useCallback(() => {
        if (active && slideShowActive) {
          if (!slideItemDuration) {
            return;
          }
          const remainingInterval =
            slideItemDuration - slideShowElapsed.current;

          // setRemainInterval(remainingInterval);
          setPaused(false);

          if (remainingInterval > 0) {
            setupTimer(remainingInterval);
          }
        }
      }, [active, slideShowActive, slideItemDuration]);

      useEffect(() => {
        if (!slideItemDuration) {
          return;
        }
        // setup the timer
        if (active && slideShowActive) {
          setStartWidth(containerWidth.current);
          setupTimer(slideItemDuration);
        }

        // disabled autofocus on active
        if (active && hasFocus) {
          containerRef.current && containerRef.current.focus();
        }
      }, [active, slideShowActive]);

      useEffect(() => {
        if (hasFocus && active) {
          containerRef.current && containerRef.current.focus();
        }
      }, [hasFocus, active]);

      const canShowReadMore = useMemo(() => {
        return useReadMore && detailedText && !customContent;
      }, []);

      const handleMediaState = useCallback(
        (state: MediaState) => {
          if (!slideShowActive) {
            return;
          }
          if (state.playing) {
            tryHandlePauseSlideshow();
          } else if (state.paused) {
            if (paused && id && onElapsed) {
              onElapsed(id);
            }
          }
        },
        [paused, slideShowActive],
      );

      const contentClass = useMemo(
        () =>
          cls(
            active ? 'timeline-card-content active' : 'timeline-card-content ',
          ),
        [active],
      );

      const contentDetailsClass = useMemo(
        () =>
          cls(
            !showMore && !customContent && useReadMore
              ? 'show-less card-description'
              : 'card-description',
          ),
        [showMore, customContent],
      );

      const handleExpandDetails = useCallback(() => {
        if ((active && paused) || !slideShowActive) {
          setShowMore(!showMore);
          onShowMore();
        }
      }, [active, paused, slideShowActive, showMore]);

      const triangleDir = useMemo(() => {
        if (flip) {
          if (branchDir === 'right') {
            return 'left';
          } else {
            return 'right';
          }
        }
        return branchDir;
      }, [branchDir, flip]);

      const canShowTriangleIcon = useMemo(() => {
        return (['VERTICAL', 'VERTICAL_ALTERNATING'] as TimelineMode[]).some(
          (m) => m === mode,
        );
      }, [mode]);

      return (
        <TimelineItemContentWrapper
          className={contentClass}
          minHeight={cardHeight}
          maxWidth={cardWidth}
          mode={mode}
          noMedia={!media}
          onClick={(ev: React.MouseEvent) => {
            ev.stopPropagation();
            if (
              !slideShowActive &&
              onClick &&
              id &&
              !disableAutoScrollOnClick
            ) {
              onClick(id);
            }
          }}
          onMouseEnter={tryHandlePauseSlideshow}
          onMouseLeave={tryHandleResumeSlideshow}
          ref={onContainerRef}
          tabIndex={0}
          theme={theme}
          borderLess={borderLessCards}
        >
          <TimelineCardHeader>
            {/* main title */}
            {!media && <MemoTitle title={title} theme={theme} url={url} />}
            {/* main timeline text */}
            {!media && <MemoSubTitle content={content} theme={theme} />}
          </TimelineCardHeader>

          {/* render media video or image */}
          {media && (
            <CardMedia
              active={active}
              cardHeight={cardHeight}
              content={content}
              hideMedia={showMore}
              id={id}
              media={media}
              onMediaStateChange={handleMediaState}
              slideshowActive={slideShowActive}
              theme={theme}
              title={title}
            />
          )}

          {/* detailed text */}
          <TimelineContentDetailsWrapper
            aria-expanded={showMore}
            className={contentDetailsClass}
            customContent={!!customContent}
            ref={detailsRef}
            theme={theme}
            useReadMore={useReadMore}
            borderLess={borderLessCards}
          >
            {customContent ? (
              <>{customContent}</>
            ) : (
              <TimelineContentDetails
                className={showMore ? 'active' : ''}
                ref={detailsRef}
                theme={theme}
              >
                {Array.isArray(detailedText)
                  ? detailedText.map((text, index) => (
                      <TimelineSubContent key={index}>
                        {text}
                      </TimelineSubContent>
                    ))
                  : detailedText}
              </TimelineContentDetails>
            )}
          </TimelineContentDetailsWrapper>

          {/* display the show more button for textual content */}
          {canShowReadMore && textContentLarge ? (
            <ShowMore
              className="show-more"
              onClick={handleExpandDetails}
              onKeyPress={(event) => {
                if (event.key === 'Enter') {
                  handleExpandDetails();
                }
              }}
              role="button"
              show={canShowMore}
              theme={theme}
              tabIndex={0}
            >
              {<span>{showMore ? 'read less' : 'read more'}</span>}
              <ChevronIconWrapper collapsed={!showMore}>
                <ChevronIcon />
              </ChevronIconWrapper>
            </ShowMore>
          ) : null}

          {canShowProgressBar && (
            <SlideShowProgressBar
              color={theme && theme.primary}
              duration={remainInterval}
              paused={paused}
              ref={progressRef}
              startWidth={startWidth}
            ></SlideShowProgressBar>
          )}

          {canShowTriangleIcon && (
            <TriangleIconWrapper
              dir={triangleDir}
              theme={theme}
            ></TriangleIconWrapper>
          )}
        </TimelineItemContentWrapper>
      );
    },
  );

TimelineCardContent.displayName = 'TimelineCardContent';

export default TimelineCardContent;
