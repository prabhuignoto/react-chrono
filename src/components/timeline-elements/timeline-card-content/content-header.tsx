import { FunctionComponent, memo, useContext, useMemo } from 'react';
import { GlobalContext } from '../../GlobalContext';
import { SubTitleMemo } from '../memoized/subtitle-memo';
import { TitleMemo } from '../memoized/title-memo';
import { ContentHeaderProps } from './header-footer.model';
import { CardTitle, TimelineCardHeader } from './timeline-card-content.styles';

/**
 * ContentHeader component
 * This component renders the header of the timeline card including the title and subtitle.
 * It doesn't render the title and subtitle if the card has media.
 * The title and subtitle are memoized to prevent unnecessary re-renders.
 *
 * @property {string} title - The title of the card.
 * @property {string} url - The URL of the card.
 * @property {boolean} media - Indicates whether the card has media or not.
 * @property {string} content - The main content of the card.
 * @returns {JSX.Element} The ContentHeader component.
 */
const ContentHeader: FunctionComponent<ContentHeaderProps> = memo(
  ({ title, url, media, content, cardTitle }: ContentHeaderProps) => {
    // Using context to get global values
    const { fontSizes, classNames, theme, isMobile } =
      useContext(GlobalContext);

    const isNotMedia = useMemo(() => !media, [media]);

    return (
      <TimelineCardHeader>
        {/* Render title if there is no media */}

        {isMobile ? (
          <CardTitle $fontSize={'1.2rem'} theme={theme}>
            {cardTitle}
          </CardTitle>
        ) : null}

        {isNotMedia ? (
          <TitleMemo
            title={title}
            theme={theme}
            url={url}
            fontSize={fontSizes?.cardTitle}
            classString={classNames?.cardTitle}
          />
        ) : null}

        {/* Render subtitle if there is no media */}

        {isNotMedia ? (
          <SubTitleMemo
            content={content}
            theme={theme}
            fontSize={fontSizes?.cardSubtitle}
            classString={classNames?.cardSubTitle}
          />
        ) : null}
      </TimelineCardHeader>
    );
  },
);

// Setting display name for easier debugging
ContentHeader.displayName = 'ContentHeader';

export { ContentHeader };
