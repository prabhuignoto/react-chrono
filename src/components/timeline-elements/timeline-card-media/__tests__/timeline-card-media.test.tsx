import { describe } from 'vitest';
import { customRender, providerProps } from '../../../common/test';
import TimelineCardMedia from '../timeline-card-media';
import { forwardRef } from 'react';
import { TextOrContentModel } from '../../timeline-card-content/text-or-content';

describe('Timeline Card media', () => {
  // should render the component with all the props
  const refComponent = forwardRef<HTMLDivElement, TextOrContentModel>(
    (props, ref) => <div ref={ref}>ref component</div>,
  );

  refComponent.displayName = 'ref component';
  it('should render the component with all the props', async () => {
    const { getByText, getByAltText, getByRole } = customRender(
      <TimelineCardMedia
        media={{
          name: 'Image',
          source: {
            type: 'URL',
            url: 'https://picsum.photos/200/300',
          },
          type: 'IMAGE',
        }}
        detailsText={refComponent}
        id="1"
        hideMedia={false}
        onMediaStateChange={() => {}}
        title="Image title"
        content="This is another test"
        cardHeight={300}
        active
      />,
      { providerProps },
    );

    expect(getByText('Image title')).toBeInTheDocument();
    expect(getByText('This is another test')).toBeInTheDocument();
    expect(getByAltText('Image')).toBeInTheDocument();
    expect(getByRole('heading', { level: 1 })).toBeInTheDocument();
  });

  // should match the snapshot
  it('should match the snapshot ( IMAGE )', () => {
    const { container } = customRender(
      <TimelineCardMedia
        media={{
          name: 'Image',
          source: {
            type: 'URL',
            url: 'https://picsum.photos/200/300',
          },
          type: 'IMAGE',
        }}
        detailsText={refComponent}
        content="This is another test"
        id="1"
        hideMedia={false}
        onMediaStateChange={() => {}}
      />,
      { providerProps },
    );
    expect(container).toMatchSnapshot();
  });

  // should match the snapshot
  it('should match the snapshot ( VIDEO )', () => {
    const { container } = customRender(
      <TimelineCardMedia
        media={{
          name: 'Video',
          source: {
            type: 'URL',
            url: 'https://www.youtube.com/watch?v=9bZkp7q19f0',
          },
          type: 'VIDEO',
        }}
        detailsText={refComponent}
        content="This is another test"
        id="1"
        hideMedia={false}
        onMediaStateChange={() => {}}
      />,
      { providerProps },
    );
    expect(container).toMatchSnapshot();
  });

  // should render video media when type is VIDEO
  it('should render video media when type is VIDEO', () => {
    const { getByTitle } = customRender(
      <TimelineCardMedia
        media={{
          name: 'Video',
          source: {
            type: 'URL',
            url: 'https://www.youtube.com/watch?v=9bZkp7q19f0',
          },
          type: 'VIDEO',
        }}
        detailsText={refComponent}
        id="1"
        hideMedia={false}
        onMediaStateChange={() => {}}
      />,
      { providerProps },
    );
    expect(getByTitle('Video')).toBeInTheDocument();
  });

  //should render the card media in text overlay mode
  it('should render the card media in text overlay mode', () => {
    const customProviderProps = {
      ...providerProps,
      textOverlay: true,
    };

    const { getByRole, getByText, getByAltText } = customRender(
      <TimelineCardMedia
        media={{
          name: 'Image',
          source: {
            type: 'URL',
            url: 'https://picsum.photos/200/300',
          },
          type: 'IMAGE',
        }}
        detailsText={refComponent}
        id="1"
        content="This is another test"
        title="Image title"
        hideMedia={false}
        onMediaStateChange={() => {}}
      />,
      { providerProps: customProviderProps },
    );

    expect(getByText('Image title')).toBeInTheDocument();
    expect(getByText('This is another test')).toBeInTheDocument();
    expect(getByAltText('Image')).toBeInTheDocument();
    expect(getByRole('heading', { level: 1 })).toBeInTheDocument();
  });

  // should render custom content in the card media
  it('should render custom content in the card media', () => {
    const { getByText } = customRender(
      <TimelineCardMedia
        media={{
          name: 'Image',
          source: {
            type: 'URL',
            url: 'https://picsum.photos/200/300',
          },
          type: 'IMAGE',
        }}
        detailsText={refComponent}
        id="1"
        hideMedia={false}
        onMediaStateChange={() => {}}
        content={<div>Custom content</div>}
      />,
      { providerProps },
    );

    expect(getByText('Custom content')).toBeInTheDocument();
  });

  // generate large chunks of random text

  // should render detailsText as expected IN HORIZONTAL MODE
  it('should render detailsText as expected', () => {
    const { getByText, getByRole } = customRender(
      <TimelineCardMedia
        media={{
          name: 'Image',
          source: {
            type: 'URL',
            url: 'https://picsum.photos/200/300',
          },
          type: 'IMAGE',
        }}
        detailsText={refComponent}
        id="1"
        hideMedia={false}
        onMediaStateChange={() => {}}
        title="Image title"
        content="This is another test"
      />,
      { providerProps: { ...providerProps, mode: 'HORIZONTAL' } },
    );

    expect(getByRole('heading', { level: 1 })).toBeInTheDocument();
    expect(getByText('Image title')).toBeInTheDocument();
  });

  // should render the arrow (triangleDir) as expected
  it('should render the arrow (triangleDir) as expected', () => {
    const { getByTestId } = customRender(
      <TimelineCardMedia
        media={{
          name: 'Image',
          source: {
            type: 'URL',
            url: 'https://picsum.photos/200/300',
          },
          type: 'IMAGE',
        }}
        detailsText={refComponent}
        id="1"
        hideMedia={false}
        onMediaStateChange={() => {}}
      />,
      { providerProps: { ...providerProps, textOverlay: true } },
    );
    expect(getByTestId('arrow-icon')).toBeInTheDocument();
  });
});
