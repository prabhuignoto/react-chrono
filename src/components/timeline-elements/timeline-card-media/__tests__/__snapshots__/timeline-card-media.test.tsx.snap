// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Timeline Card media > should match the snapshot ( IMAGE ) 1`] = `
<div>
  <div
    class="sc-bgqQcB eKaMuE card-media-wrapper card-media"
    mode="VERTICAL_ALTERNATING"
  >
    <div
      class="sc-ispOId hMrjPH"
      height="200"
    >
      <img
        alt="Image"
        class="sc-gTRrQi fvRpJr"
        loading="lazy"
        mode="VERTICAL_ALTERNATING"
        role="img"
        src="https://picsum.photos/200/300"
      />
    </div>
  </div>
  <div
    class="sc-fFGjHI kHvLaM"
    mode="VERTICAL_ALTERNATING"
  >
    <div
      class="sc-kPTPQs iZPJSj"
    />
    <h2
      class="sc-kuWgmH lnojIa card-sub-title card-subtitle"
      font-size="0.85rem"
    >
      This is another test
    </h2>
  </div>
</div>
`;

exports[`Timeline Card media > should match the snapshot ( VIDEO ) 1`] = `
<div>
  <div
    class="sc-bgqQcB eKaMuE card-media-wrapper card-media"
    mode="VERTICAL_ALTERNATING"
  >
    <iframe
      allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
      allowfullscreen=""
      class="sc-iVCKna jrpgms"
      frameborder="0"
      src="https://www.youtube.com/watch?v=9bZkp7q19f0?enablejsapi=1"
      title="Video"
    />
  </div>
  <div
    class="sc-fFGjHI kHvLaM"
    mode="VERTICAL_ALTERNATING"
  >
    <div
      class="sc-kPTPQs iZPJSj"
    />
    <h2
      class="sc-kuWgmH lnojIa card-sub-title card-subtitle"
      font-size="0.85rem"
    >
      This is another test
    </h2>
  </div>
</div>
`;
